src/eval.ml

```diff
@@ -218,11 +258,92 @@ and eq_modulo : term -> term -> bool = fun a b ->
   let res = try eq_modulo [(a,b)]; true with Exit -> false in
   if !log_enabled then log_eqmd (r_or_g res "%a == %a") pp a pp b; res
 
+(** [tree_walk t c s] tries to match stack [s] against tree [t] of capacity
+    [c]. *)
+and tree_walk : Dtree.t -> int -> term list -> term option =
+  fun tree capa stk ->
+  let vars = Array.make capa Kind in (* dummy terms *)
+  let module R = Dtree.ReductionStack in
+  let stk = R.of_list stk in
+  (* [walk t s c] where [s] is the stack of terms to match and [c] the cursor
+     indicating where to write in the [env] array described in {!module:Terms}
+     as the environment of the RHS during matching. *)
+  let rec walk tree stk cursor =
+    match tree with
+    | Fail                                   -> None
+    | Leaf(env_builder, act)                 ->
+        (* Allocate an environment for the action. *)
+        let env = Array.make (Bindlib.mbinder_arity act) TE_None in
+        (* Retrieve terms needed in the action from the [vars] array. *)
+        let fn pos slot =
+          let t = unfold vars.(pos) in
+          let b = Bindlib.raw_mbinder [||] [||] 0 mkfree (fun _ -> t) in
+          env.(slot) <- TE_Some(b)
+        in
+        IntMap.iter fn env_builder;
+        (* Actually perform the action. *)
+        Some(Bindlib.msubst act env)
+    | Node({swap; children; store; default}) ->
+        (* Quit if stack is too short. *)
+        if R.is_empty stk || swap >= R.length stk then None else
+        (* Pick the right term in the stack. *)
+        let (left, examined, right) = R.destruct stk swap in
```

Comment:
destruct can raise Not_found and Not_found is not caught here. Please add a comment if this cannot happen.